sdks/python/apache_beam/io/gcp/gcsio.py

```diff
@@ -456,364 +465,82 @@ def size_of_files_in_glob(self, pattern, limit=None):
-    self.mode = mode
-    self.buffer_size = buffer_size
+class GcsDownloader(Downloader):
+  def __init__(self, client, path, buffer_size):
+    self._client = client
+    self._path = path
+    self._bucket, self._name = parse_gcs_path(path)
+    self._buffer_size = buffer_size
 
     # Get object state.
-    self.get_request = (storage.StorageObjectsGetRequest(
-        bucket=self.bucket, object=self.name))
+    self._get_request = (storage.StorageObjectsGetRequest(
+        bucket=self._bucket, object=self._name))
     try:
-      metadata = self._get_object_metadata(self.get_request)
+      metadata = self._get_object_metadata(self._get_request)
     except HttpError as http_error:
       if http_error.status_code == 404:
-        raise IOError(errno.ENOENT, 'Not found: %s' % self.path)
+        raise IOError(errno.ENOENT, 'Not found: %s' % self._path)
       else:
-        logging.error('HTTP error while requesting file %s: %s', self.path,
+        logging.error('HTTP error while requesting file %s: %s', self._path,
                       http_error)
         raise
-    self.size = metadata.size
+    self._size = metadata.size
 
     # Ensure read is from file of the correct generation.
-    self.get_request.generation = metadata.generation
+    self._get_request.generation = metadata.generation
 
     # Initialize read buffer state.
     self.download_stream = cStringIO.StringIO()
-    self.downloader = transfer.Download(
-        self.download_stream, auto_transfer=False, chunksize=self.buffer_size)
-    self.client.objects.Get(self.get_request, download=self.downloader)
-    self.position = 0
-    self.buffer = ''
-    self.buffer_start_position = 0
-    self.closed = False
+    self._downloader = transfer.Download(
+        self.download_stream, auto_transfer=False, chunksize=self._buffer_size)
+    self._client.objects.Get(self._get_request, download=self._downloader)
 
   @retry.with_exponential_backoff(
       retry_filter=retry.retry_on_server_errors_and_timeout_filter)
   def _get_object_metadata(self, get_request):
-    return self.client.objects.Get(get_request)
-
-  def __iter__(self):
-    return self
-
-  def __next__(self):
-    """Read one line delimited by '\\n' from the file.
-    """
-    return next(self)
-
-  def next(self):
-    """Read one line delimited by '\\n' from the file.
-    """
-    line = self.readline()
-    if not line:
-      raise StopIteration
-    return line
-
-  def read(self, size=-1):
-    """Read data from a GCS file.
-
-    Args:
-      size: Number of bytes to read. Actual number of bytes read is always
-            equal to size unless EOF is reached. If size is negative or
-            unspecified, read the entire file.
-
-    Returns:
-      data read as str.
-
-    Raises:
-      IOError: When this buffer is closed.
-    """
-    return self._read_inner(size=size, readline=False)
+    return self._client.objects.Get(get_request)
 
-  def readline(self, size=-1):
```

Comment:
<!--new_thread; commit:71192743121d0b142828da870905369f15c64f15; resolved:0-->
There may a subtle behavior change now that we remove this custom `readline()` code.  The standard library code (`_pyio.py`) may do some `\r\n` -> `\n` translation depending on the read mode and platform, which we don't want.  Can you verify that this won't be the case, even on Windows?

CC: @chamikaramj

Reply:
TextIOWrapper is not used, thus no translation is performed and readline() uses b"\n" as a line terminator.

Reply:
Thanks!